### PR TITLE
Test exclude

### PR DESCRIPTION
**Pull request by matthewfarwell**
Added exclusions to testOnly and testQuick command line, indicated by a '-' prefix.

### CODE REVIEW

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

File: main/src/main/scala/sbt/Defaults.scala [hunk 1]
```diff
@@ -566,11 +566,18 @@ object Defaults extends BuildCommon {
...
+
+      val includeFilters = includeArgs map GlobFilter.apply
+      val excludeFilters = excludeArgs.map(_.substring(1)).map(GlobFilter.apply)
+
+      if (includeFilters.isEmpty && excludeArgs.isEmpty)
         Seq(const(true))
-      else
-        filters.map { f => (s: String) => f accept s }
+      else if (includeFilters.isEmpty) {
+        excludeFilters.map { f => (s: String) => !f.accept(s) }
```
Comment (jsuereth) [inline main/src/main/scala/sbt/Defaults.scala]: I'm not sure this works the way you want it to.  Are all these filters "summed" up at the end, or "productized"?

i.e. you may want a single function that takes the string, and then runs through all the excludes, like:

``` scala
Seq( { (s: String) => excludeFilters forall (f => !(f accept s)) })
```

File: sbt/src/sbt-test/tests/test-exclude/test [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# with only exclusion
+> test-only -Test1
+-$ exists target/Test1.run
+$ exists target/Test2.run
+
+$ delete target/Test2.run
+
+
+# with only glob exclusion
+> test-only -Test*
```
Comment (jsuereth) [inline sbt/src/sbt-test/tests/test-exclude/test:53]: Here, we should have an additional test with some arbitrary string, like -Blah to see if my intuitation about joining filters is correct.

**Comment by jsuereth** [comment]
One minor comment on how we join pure exclusion filters, but otherwise LGTM.

cc> @eed3si9n ?

File: sbt/src/sbt-test/tests/test-exclude/test (hunk 2 above)
Comment (matthewfarwell) [inline sbt/src/sbt-test/tests/test-exclude/test:53]: Done.

**Comment by matthewfarwell** [comment]
You are right. I corrected the logic and added more tests (unit & scripted).

**Comment by DavidPerezIngeniero** [comment]
Which version of SBT will have this available?
Maybe 0.13.9?

**Comment by jsuereth** [comment]
Cool, LGTM.  We can include in sbt 0.13.9, I _think_ our tests are fixed, so let's get this through :)

**Comment by jsuereth** [comment]
Oh, this is  based on the flaky maven-test base, so we can merge as-is.

File: main/src/main/scala/sbt/Defaults.scala [hunk 3]
```diff
@@ -892,7 +901,7 @@ object Defaults extends BuildCommon {
       selectTests ~ options
     }
 
-  def distinctParser(exs: Set[String], raw: Boolean): Parser[Seq[String]] =
+  private def distinctParser(exs: Set[String], raw: Boolean): Parser[Seq[String]] =
```
Comment (dwijnand) [inline main/src/main/scala/sbt/Defaults.scala:904]: We missed this. This is a breaking change, see https://travis-ci.org/typesafehub/zinc/builds/79476798

```
[error] /home/travis/build/typesafehub/zinc/project/Scriptit.scala:56: method distinctParser in object Defaults cannot be accessed in object sbt.Defaults
[error]     Defaults.distinctParser(scriptitTests(scriptitBase, script).toSet, raw = false)
[error]              ^
```

**Pull request merged**